tangelo/toolboxes/circuits/lcu.py

```diff
@@ -0,0 +1,310 @@
+    """
+
+    # Incorporate sign of time into qubit operator.
+    qu_op = np.sign(t_in) * qu_op_in
+    t = abs(t_in)
+
+    # remove the coefficient value and generate the list of QubitOperator with only its phase 1j, or -1j
+    vector = list()
+    unitaries = list()
+    for term, coeff in qu_op.terms.items():
+        if np.abs(coeff.imag) > 1.e-7:
+            raise ValueError(f"Only real qubit operators are allowed but term {term} has coefficient {coeff}")
+        if coeff.real > 0:
+            unitaries.append(QubitOperator(term, -1j))
+        else:
+            unitaries.append(QubitOperator(term, 1j))
+        vector.append(np.abs(coeff.real))
+
+    num_terms = len(vector)
+    vector = np.array(vector)
+    vsum = sum(vector)
+
+    # Calculate 1-norm of coefficients in qubit operator and obtain the maximum time-step allowed
+    # These values are obtained by finding the relevant root of the kth order Taylor series polynomial approximation
+    # of 2 = exp(x) (i.e. the roots of 2 = \sum_{k=0}^N x^k/k!). The limit for large k is log(2)
+    poly_roots = {1: 1., 2: 0.73205081, 3: 0.69888549, 4: 0.69390315, 5: 0.69323260, 6: 0.69315552, 7: 0.69314790, 8: 0.69314724}
+    max_time_step = poly_roots[kmax] / vsum if kmax < 9 else np.log(2)
+
+    # Calculate the number of time steps required and calculate the actual 1-norm for each time-step
+    time_steps = math.ceil(t / max_time_step)
+    vsum_with_t = sum(coeff * t / time_steps for coeff in vector)
+    expvsum = sum((vsum_with_t) ** k / math.factorial(k) for k in range(0, kmax+1))
+
+    # Generate vector v_i = sqrt(alpha_i)/np.sqrt(1-norm) with zeros padded to create a vector of length 2**n
+    vector = np.sqrt(vector) / np.sqrt(vsum)
+    n_qubits = math.ceil(math.log2(num_terms))
+    newvec = np.zeros(2 ** n_qubits)
+    newvec[:num_terms] = vector
+
+    # Calculate circuit that generates vector and apply controls for Taylor series
+    s = StateVector(newvec, order="msq_first")
+    qss = list()
+    for k in range(kmax):
+        qss.append(s.initializing_circuit())
+        # Reindex to correct block of qubits for kth order of Taylor series
+        qss[-1].reindex_qubits(list(range(kmax + k * n_qubits, kmax + (k + 1) * n_qubits)))
+        # Add control for the kth unary encoded qubits
+        for gate in qss[-1]._gates:
+            if gate.control is not None:
+                gate.control += [k]
+            else:
+                gate.name = "C"+gate.name
+                gate.control = [k]
+    qstot = Circuit()
+    for qs in qss:
+        qstot += qs
+
+    # Calculate coefficients for unary encoding k="0"+"0"*(kmax-k)*"1"*k where
+    # first qubit encodes extra identity term needed to ensure 1-norm = 2 such that
+    # oblivious amplitude amplification is applicable
+    kvec = np.zeros(2**(kmax + 1))
+    for k in range(0, kmax+1):
+        pos = int("0" + "0" * (kmax - k) + "1" * k, base=2)
```

Comment:
For readability, its useful to vary the number of spaces. Spacing everything equally (whether no space of 1 space) makes it almost equally hard to figure out operation priority. Here's my personal preference here, for example; it helps me see blocks more easily.

`pos = int("0" + "0"*(kmax-k) + "1"*k, base=2)`